src/profiler.cpp

```diff
@@ -1647,6 +1652,97 @@ void Profiler::dumpText(Writer& out, Arguments& args) {
+        std::vector<std::string>& values, const std::string& value) {
+    const auto& pair = idx_map.insert({value, idx_map.size()});
+    if (pair.second) {
+        values.push_back(value);
+    }
+    return pair.first->second;
+}
+
+void Profiler::dumpOtlp(Writer& out, Arguments& args) {
+    _otlp_buffer.reset();
+
+    std::unordered_map<std::string, u64> string_idx_map;
+    std::vector<std::string> strings_vec;
+    std::unordered_map<std::string, u64> function_idx_map;
+    std::vector<std::string> functions_vec;
+
+    protobuf_mark_t scope_profiles_mark = _otlp_buffer.startMessage(Otlp::ResourceProfiles::scope_profiles);
+    protobuf_mark_t profile_mark = _otlp_buffer.startMessage(Otlp::ScopeProfiles::profiles);
+
+    // TODO: Include per-binary debug info
+    protobuf_mark_t mapping_mark = _otlp_buffer.startMessage(Otlp::Profile::mapping_table);
+    _otlp_buffer.commitMessage(mapping_mark);
+
+    _otlp_buffer.field(Otlp::Profile::period, (u64) args._interval);
+
+    protobuf_mark_t sample_type_mark = _otlp_buffer.startMessage(Otlp::Profile::sample_type);
+    // TODO: This should be changed when other profling modes (e.g. alloc), or --total are used
+    _otlp_buffer.field(Otlp::ValueType::type_strindex, maybeAddToIdxMap(string_idx_map, strings_vec, "samples"));
+    _otlp_buffer.field(Otlp::ValueType::unit_strindex, maybeAddToIdxMap(string_idx_map, strings_vec, "count"));
+    _otlp_buffer.field(Otlp::ValueType::aggregation_temporality, Otlp::AggregationTemporality::delta);
+    _otlp_buffer.commitMessage(sample_type_mark);
+
+    protobuf_mark_t period_type_mark = _otlp_buffer.startMessage(Otlp::Profile::period_type);
+    // TODO: This should be changed when other profling modes (e.g. alloc), or --total are used
+    _otlp_buffer.field(Otlp::ValueType::type_strindex, maybeAddToIdxMap(string_idx_map, strings_vec, "cpu"));
+    _otlp_buffer.field(Otlp::ValueType::unit_strindex, maybeAddToIdxMap(string_idx_map, strings_vec, "nanoseconds"));
+    _otlp_buffer.commitMessage(period_type_mark);
+
+    std::vector<CallTraceSample*> samples;
+    _call_trace_storage.collectSamples(samples);
+
+    u64 frames_seen = 0;
+    FrameName fn(args, args._style & ~STYLE_ANNOTATE, _epoch, _thread_names_lock, _thread_names);
+    for (auto it = samples.begin(); it != samples.end(); ++it) {
+        CallTrace* trace = (*it)->acquireTrace();
+        if (trace == NULL || excludeTrace(&fn, trace)) continue;
+
+        u64 counter = args._counter == COUNTER_SAMPLES ? (*it)->samples : (*it)->counter;
+        if (counter == 0) continue;
+
+        u64 num_frames = trace->num_frames;
+        for (u64 j = 0; j < num_frames; ++j) {
+            u64 function_idx = maybeAddToIdxMap(function_idx_map, functions_vec, fn.name(trace->frames[j]));
+            _otlp_buffer.field(Otlp::Profile::location_indices, function_idx);
+        }
+
+        protobuf_mark_t sample_mark = _otlp_buffer.startMessage(Otlp::Profile::sample);
```

Comment:
Your proto writer does not seem to benefit from `repeated` fields - it always writes header for each single message.

Reply:
Right, I always use the `expanded` mode for both primitive and LEN-type values: https://protobuf.dev/editions/features/#repeated_field_encoding

This won't pose any problem on the reader side, but it's definitely something we could improve in the future. I didn't want to overcomplicate the interface, but I'm keeping track of this and other future improvements (like the optimistic encoding of nested messages which I proposed [here](https://github.com/async-profiler/async-profiler/pull/1292#discussion_r2088484573))